Refactored Component::extendRealizeTopology to use for range loops

adamkewley opened the pull request:
Tiny Friday-afternoon refactor of a code section I worked near when implementing the new `CacheVariable<T>` abstraction.

No functional changes: it should behave like-for-like. I've deliberately kept this small so that, once it's through CI, it can just be quickly checked + merged.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/opensim-org/opensim-core/2875)
<!-- Reviewable:end -->

aymanhab reviewed (approved):
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2875)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

aymanhab reviewed (approved):
LGTM @adamkewley just curious why we ever needed const_cast but not anymore.

> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2875#-:-MGy22cSDP3WgP_q6xQN:b-fb380a)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

adamkewley commented:
@aymanhab good question. I had 5 minutes to spare and I love archeology :D

I looked into this thinking "what idiot put that `const_cast` there. Little did I know, but it was I was the murderer all along:

```bash
git blame Component.cpp | grep "const_cast"
# others
b5fbf0cd91 (Adam Kewley            2020-06-18 12:33:34 +0100 1414)     Component *mutableThis = const_cast<Component*>(this);
b5fbf0cd91 (Adam Kewley            2020-06-18 12:33:34 +0100 1441)                 AddedStateVariable* masv = const_cast<AddedStateVariable*>(asv);
```

But the commit that introduced it was actually dragged in from early `CacheVariable<T>` work, where I refactored this function to use range loops and remove the const cast (essentially what this PR is now doing separately) in June. I didn't want the headache of a massive diff for the `CacheVariable<T>` work, which would drag out the PR unnecessarily, so I created a revert commit that re-introduced the blight from upstream:

```diff
commit b5fbf0cd91fde054e73dfabe97c27dadc7aa7e8d
Author: Adam Kewley <contact@adamkewley.com>
Date:   Thu Jun 18 12:33:34 2020 +0100

    Reverted some of the crud from upstream to make this changeset smaller

diff --git a/OpenSim/Common/Component.cpp b/OpenSim/Common/Component.cpp
index fb146ebb9..d6d15c2c3 100644
--- a/OpenSim/Common/Component.cpp
+++ b/OpenSim/Common/Component.cpp
@@ -1375,33 +1375,54 @@ getStateVariableNamesAddedByComponent() const
 void Component::extendRealizeTopology(SimTK::State& s) const
 {
     const SimTK::Subsystem& subSys = getSystem().getDefaultSubsystem();
+    
+    Component *mutableThis = const_cast<Component*>(this);
 
     // Allocate Modeling Option
-    for (auto& p : this->_namedModelingOptionInfo) {
-        ModelingOptionInfo& moi = p.second;
-        moi.index = subSys.allocateDiscreteVariable(s, SimTK::Stage::Instance, new SimTK::Value<int>(0));
+    if(_namedModelingOptionInfo.size()>0){
(etc)
```

And the original refactoring pass, pre `CacheVariable<T>` implementation, was 4 commits earlier (`b5fbf~4`), or `9e67`:

```diff
diff --git a/OpenSim/Common/Component.h b/OpenSim/Common/Component.h
commit 9e67a2aa8d8327e3dc1a8f35956281389bd1a304
Author: Adam Kewley <contact@adamkewley.com>
Date:   Mon Jun 15 16:02:23 2020 +0100

    Further refactoring and clarifications (compiles + works)

(...)
-    
-    Component *mutableThis = const_cast<Component*>(this);
 
     // Allocate Modeling Option
-    if(_namedModelingOptionInfo.size()>0){
-        std::map<std::string, ModelingOptionInfo>::iterator it;
-        for (it = (mutableThis->_namedModelingOptionInfo).begin(); 
-             it !=_namedModelingOptionInfo.end(); ++it)
-        {
-            ModelingOptionInfo& moi = it->second;
-            moi.index = subSys.allocateDiscreteVariable
-               (s, SimTK::Stage::Instance, new SimTK::Value<int>(0));
-        }
+    for (auto& p : this->_namedModelingOptionInfo) {
+        ModelingOptionInfo& moi = p.second;
+        moi.index = subSys.allocateDiscreteVariable(s, SimTK::Stage::Instance, new SimTK::Value<int>(0));
     }
```

So it was in `b5fbf~5`:

```
git blame Component.cpp b5fbf~5 | grep "const_cast"
(...)
2da6e1d5f3 (Ajay Seth              2014-01-21 20:26:59 +0000 1388)     Component *mutableThis = const_cast<Component*>(this);
```

So Ajay was who poisoned the wine in 2014, but why?! 

```
commit 2da6e1d5f3020019c01a717eddbc81c771875884
Author: Ajay Seth <aseth@stanford.edu>
Date:   Tue Jan 21 20:26:59 2014 +0000

    First check-in of the base Component class...
```

Original sin :D

adamkewley commented:
"Original sin" doesn't adequately explain it, but I think *actual* reason is because the original implementation needs this line to compile:

```
ModelingOptionInfo& moi = it->second;
```

Which wouldn't work for a `const` iterator, which is the default if `_namedModelingOptionInfo` transitively inherits `*this`'s `const`ness (any member should). But what probably has happened is that, at some point, that member was given a `mutable` qualifier, which lets downstream users of the member scrap the `const` qualifer. 

Probably not be the reason though. The `mutable` qualifier has been in there since day one, so it's more likely that the `const_cast` was added because the original `for` loop's `.begin()` call was misbehaving (and always returning a `const` iterator) whereas a `for`-range loop kind of hides whatever magic is necessary to get a non-`const` iterator.

---

This dive turned into 15 minutes but, again, it's always a nice detour to investigate why some patterns appear in codebases, etc. - especially larger ones like OpenSim.

I'll merge this refactor and prep for the next one! :)

aymanhab commented:
I feel guilty now sending you on this chase, but nice job digging. Thank you 👍

adamkewley commented:
No no, I'm not being sarcastic or anything @aymanhab - I needed a break and my `git`-fu was overdue for a little exercise :)
